test/controllers/IncapsulaControllerTest.ts

```diff
@@ -0,0 +1,481 @@
+import * as Lab from '@hapi/lab';
+import * as nock from 'nock';
+import * as request from 'supertest';
+import { server } from '../../src/darkstar';
+export const lab = Lab.script();
+const { describe, it, beforeEach } = lab;
+
+describe('/v1/caches/incapsula', () => {
+  let incapsulaAPIMock: nock.Scope;
+
+  beforeEach(() => {
+    incapsulaAPIMock = nock('https://my.incapsula.com');
+  });
+
+  describe('/zones/${zone_id}', () => {
+    describe('DELETE', () => {
+      let flushRequest: request.Test;
+      let incapsulaFlushMock: nock.Interceptor;
+
+      beforeEach(() => {
+        flushRequest = request(server.listener)
+          .delete('/v1/caches/incapsula/zones/abcd')
+          .set('accept', 'application/json');
+        incapsulaFlushMock = incapsulaAPIMock.post(
+          '/api/prov/v1/sites/cache/purge?api_id=1234&api_key=4321&site_id=abcd'
+        );
+      });
+
+      it('should flush a complete zone of the Incapsula cache', () => {
+        incapsulaFlushMock.matchHeader('accept', 'application/json').reply(200, { status: 'ok' });
+
+        return new Promise((resolve, reject) => {
+          flushRequest
+            .send({
+              incapsulaApiID: '1234',
+              incapsulaApiKey: '4321',
+            })
+            .expect('content-type', /application\/json/)
+            .expect(200)
+            .expect({
+              remoteStatusCode: 200,
+              remoteResponse: { status: 'ok' },
+            })
+            .end((error: any, _: request.Response) => {
+              if (error) {
+                console.log('TCL: error', error);
+                reject(error);
+              }
+              incapsulaAPIMock.done();
+              resolve();
+            });
+        });
+      });
+
+      it('should reply "Bad request" when invalid payload is sent', () => {
+        incapsulaFlushMock.times(0);
+
+        return new Promise((resolve, reject) => {
+          flushRequest
+            .send({})
+            .expect('content-type', /application\/json/)
+            .expect(400)
+            .expect({
+              message: 'child "incapsulaApiID" fails because ["incapsulaApiID" is required]',
+              validation: { source: 'payload', keys: ['incapsulaApiID'] },
+            })
+            .end((error: any, _: request.Response) => {
+              if (error) {
+                reject(error);
+              }
+              incapsulaAPIMock.done();
+              resolve();
+            });
+        });
+      });
+
+      it('should reply bad request when receiving Incapsula cache client errors', () => {
+        const incapsulaError = {
```

Comment:
This is not an Incapsula error message, see https://docs.imperva.com/bundle/cloud-application-security/page/api/api.htm#General2